pkg/receive/hashring.go

```diff
@@ -108,7 +119,53 @@ func NewHashring(matcher Matcher, groups []*targetgroup.Group) Hashring {
 		matcher:   matcher,
 	}
 	for _, g := range groups {
-		m.hashrings[g.Source] = &simpleHashring{*g}
+		l, ok := g.Labels[tenantLabel]
+		if !ok {
+			l = ""
+		}
+		t := string(l)
+		if _, ok := m.hashrings[t]; !ok {
+			m.hashrings[t] = &simpleHashring{}
+		}
+		h := m.hashrings[t].(*simpleHashring)
+		h.Targets = append(h.Targets, g.Targets...)
 	}
 	return m
 }
+
+// HashringsAreEquivalent indicates whether the hashrings created from
+// two sets of groups would be equivalent.
+func HashringsAreEquivalent(a, b []*targetgroup.Group) bool {
```

Comment:
In the comment for this function. Could you give a quick example on why we actually need this? Like an (example) use case.

Reply:
It is used here: https://github.com/squat/thanos/blob/df3289e9e914846d08bd28987a368aac6083f137/cmd/thanos/receive.go#L152

This is required so we know if the hash ring configuration has changed. The way file sd works, every interval seconds, the file is re-read and an update is sent even if the file hasn’t changed. This should not cause the receiver to shut down because the configuration is the same. For that reason we need to check if two file sd updates are equivalent. If they are, we can ignore the update. If they are different, we shutdown gracefully.

Reply:
I can summarize that in the comment